sys/random/fortuna.c

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (C) 2016-2018 Bas Stottelaar <basstottelaar@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sys_random
+ * @{
+ * @file
+ *
+ * @brief   Glue-code for Fortuna PRNG.
+ *
+ * @author  Bas Stottelaar <basstottelaar@gmail.com>
+ * @}
+ */
+
+#include "log.h"
+#include "mutex.h"
+
+#include "fortuna/fortuna.h"
+
+/**
+ * @brief This holds the PRNG state.
+ */
+static fortuna_state_t fortuna_state;
+
+/**
+ * @brief Initialize the PRNG with a given number of bytes.
+ */
+static void _init(uint8_t *in, size_t bytes)
+{
+    fortuna_seed_t seed;
+
+    /* ensure a seed of proper length is available, which may not be the case
+       with a small input */
+    memset(seed, 0, sizeof(seed));
+
+    for (int i = 0; i < (int) bytes; i++) {
+        seed[i % sizeof(seed)] ^= in[i];
+    }
+
+    /* initialize the PRNG state */
+    fortuna_init(&fortuna_state);
+
+    /* update the PRNG seed (seed will be overwritten by design!) */
+    fortuna_update_seed(&fortuna_state, &seed);
+}
+
+/**
+ * @brief Wrapper for fortuna_random_data that supports reading more than
+ *        FORTUNA_RESEED_LIMIT and also checks the result.
+ */
+static void _read(uint8_t *out, size_t bytes)
+{
+    do {
+        /* read at most chunk bytes to not exhaust the state at once */
+        size_t chunk = bytes < FORTUNA_RESEED_LIMIT ?
+                       bytes : FORTUNA_RESEED_LIMIT;
+
+        int res = fortuna_random_data(&fortuna_state, out, bytes);
```

Comment:
shouldn't this use `chunk` instead of `bytes`, i.e., `fortuna_random_data(&fortuna_state, out, chunk)`?

Reply:
Yes, good catch!